playbooks/tasks/portal-docker-services-check-running.yml

```diff
@@ -0,0 +1,123 @@
+---
+# Check docker services are running (not restarting)
+# Inputs:
+# - docker_containers_check_list: List of docker container names to check
+# - timeout_secs: Total timeout in seconds to wait for containers running
+#     successfully, not restarting (expected to be 120+ seconds)
+
+- name: Assert that inputs are defined correctly
+  ansible.builtin.assert:
+    that:
+      - docker_containers_check_list is defined
+      - docker_containers_check_list | length > 0
+      - docker_containers_check_timeout_secs is defined
+      - docker_containers_check_timeout_secs > 0
+    fail_msg: |
+      Variable docker_containers_check_list must be defined
+      and it must be a list containing at least 1 docker service.
+
+      Variable docker_containers_check_timeout_secs must be defined and greater
+      than 0. It is a playbook's total timeout that we give for all docker
+      containers to start correctly (and not restarting).
+
+- name: Set main start timestamp once
+  set_fact:
+    # We want to get local time using lookup/pipe
+    check_start_epoch: "{{ lookup('ansible.builtin.pipe', 'date +%s') }}"
+  when: >-
+    check_start_epoch is not defined
+    or check_start_epoch == 0
+
+# We want to get current check timestamp from a remote host, because docker
+# container info StartedAt timestamp is also from remote host, so we can't use
+# lookup/pipe.
+
+- name: Get current check epoch timestamp from remote host
+  ansible.builtin.command: date +%s
+  register: date_result
+
+- name: Set current check epoch timestamp from remote host
+  set_fact:
+    current_check_start_epoch: "{{ date_result.stdout }}"
+
+- name: Check docker containers status
+  community.docker.docker_container_info:
+    name: "{{ item }}"
+  register: docker_info_result
+  loop: "{{ docker_containers_check_list }}"
+
+- name: Reset docker services are running successfully flag
+  set_fact:
+    docker_services_running_ok: True
+
+- name: Set docker services are running successfully flag
+  vars:
+    min_required_non_restarting_time_in_secs: 65
+  set_fact:
+    # Check that container exists, is running and is not restarting for more
+    # more than 1 minute. We need to wait more than 1 minute, because container
+    # can seem running ok (not restating) and restart after 1 minute.
+    # Time check condition:
+    #   (current_check_start_epoch in UTF in secs) - (container started at)
+    #   > min time required
+    docker_services_running_ok: >-
+      {{
+        docker_services_running_ok
+        and item.exists
+        and item.container.State.Running
+        and not item.container.State.Restarting
+        and
+          current_check_start_epoch | int
+          -
+          (item.container.State.StartedAt[:19] | to_datetime('%Y-%m-%dT%H:%M:%S'))
+            .strftime('%s') | int
+          > min_required_non_restarting_time_in_secs
+      }}
+  loop: "{{ docker_info_result.results }}"
```

Comment:
This is going to run in parallel for every docker service right? It's not going to run for each one by one?

Reply:
It's something in between due to Ansible limitations but performance is ok.
Steps:

- `Get current check epoch timestamp from remote host`
- `Check docker containers status`
    - this step runs in serial
    - it takes total approx. 5 - 10 seconds for all containers combined
- `Set docker services are running successfully flag`
    - this step runs in serial
    - takes total ~ 500 ms (not 65 seconds for each container)
- timeout check